serde/src/de/mod.rs

```diff
@@ -230,6 +231,149 @@ pub trait Deserialize: Sized {
+/// struct ExtendVec<'a, T: 'a>(&'a mut Vec<T>);
+///
+/// impl<'a, T> DeserializeSeed for ExtendVec<'a, T>
+///     where T: Deserialize
+/// {
+///     // The return type of the `deserialize` method. This implementation
+///     // appends onto an existing vector but does not create any new data
+///     // structure, so the return type is ().
+///     type Value = ();
+///
+///     fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+///         where D: Deserializer
+///     {
+///         // Visitor implementation that will walk an inner array of the JSON
+///         // input.
+///         struct ExtendVecVisitor<'a, T: 'a>(&'a mut Vec<T>);
+///
+///         impl<'a, T> Visitor for ExtendVecVisitor<'a, T>
+///             where T: Deserialize
+///         {
+///             type Value = ();
+///
+///             fn visit_seq<V>(self, mut visitor: V) -> Result<(), V::Error>
+///                 where V: SeqVisitor
+///             {
+///                 // Visit each element in the inner array and push it onto
+///                 // the existing vector.
+///                 while let Some(elem) = visitor.visit()? {
+///                     self.0.push(elem);
+///                 }
+///                 Ok(())
+///             }
+///         }
+///
+///         deserializer.deserialize_seq(ExtendVecVisitor(self.0))
+///     }
+/// }
+///
+/// // Visitor implementation that will walk the outer array of the JSON input.
+/// struct FlattenedVecVisitor<T>(PhantomData<T>);
+///
+/// impl<T> Visitor for FlattenedVecVisitor<T>
+///     where T: Deserialize
+/// {
+///     // This Visitor constructs a single Vec<T> to hold the flattened
+///     // contents of the inner arrays.
+///     type Value = Vec<T>;
+///
+///     fn visit_seq<V>(self, mut visitor: V) -> Result<Vec<T>, V::Error>
+///         where V: SeqVisitor
+///     {
+///         // Create a single Vec to hold the flattened contents.
+///         let mut vec = Vec::new();
+///
+///         // Each iteration through this loop is one inner array.
+///         while let Some(()) = visitor.visit_seed(ExtendVec(&mut vec))? {
+///             // Nothing to do; inner array has been appended into `vec`.
+///         }
+///
+///         // Return the finished vec.
+///         Ok(vec)
+///     }
+/// }
+///
+/// # fn example<D: Deserializer>(deserializer: D) -> Result<(), D::Error> {
+/// let visitor = FlattenedVecVisitor(PhantomData);
+/// let flattened: Vec<u64> = deserializer.deserialize_seq(visitor)?;
+/// # Ok(()) }
+/// ```
+pub trait DeserializeSeed: Sized {
+    /// The type produced by using this seed.
+    type Value;
+
+    /// Equivalent to the more common `Deserialize::deserialize` method, except
+    /// with some initial piece of data (the seed) passed in.
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+        where D: Deserializer;
+}
+
+impl<T> DeserializeSeed for PhantomData<T>
```

Comment:
PhantomData is a bit magical in that it has a type parameter but you can use it like a unit struct.

```rust
pub struct PhantomData<T: ?Sized>;

pub struct NoSeed<T: ?Sized>(PhantomData<T>);
```

So you can call `f(PhantomData)` but you cannot call `f(NoSeed)`. It would need to be `f(NoSeed::new())` or `f(NoSeed(PhantomData))` or something unfortunate like that.

The workaround for this would be what we did in [serde-stateful-deserialize](https://github.com/sfackler/serde-stateful-deserialize) which is:

```rust
pub use std::marker::PhantomData as NoSeed;
```

This reexports both the type and the unit-like way of referring to it, so it lets you do `f(NoSeed)`. The downside is there is no way to override PhantomData's documentation so in our rustdoc there would be a type `serde::de::NoSeed` with all of PhantomData's documentation and impls which is confusing.

Do you think NoSeed::new() would work better?